Sources/NIOTransportServices/StateManagedNWConnectionChannel.swift

```diff
@@ -0,0 +1,617 @@
+                self.reusePort = (value as! SocketOptionValue) != Int32(0)
+            default:
+                try self.nwOptions.applyChannelOption(option: optionValue, value: value as! SocketOptionValue)
+            }
+        case _ as ChannelOptions.Types.WriteBufferWaterMarkOption:
+            if self._backpressureManager.writabilityChanges(whenUpdatingWaterMarks: value as! ChannelOptions.Types.WriteBufferWaterMark) {
+                self.pipeline.fireChannelWritabilityChanged()
+            }
+        case is NIOTSChannelOptions.Types.NIOTSEnablePeerToPeerOption:
+            self.enablePeerToPeer = value as! NIOTSChannelOptions.Types.NIOTSEnablePeerToPeerOption.Value
+        case _ as NIOTSChannelOptions.Types.NIOTSWaitForActivityOption:
+            let newValue = value as! Bool
+            self.options.waitForActivity = newValue
+
+            if let state = self.connection?.state, case .waiting(let err) = state, !newValue {
+                // We're in waiting now, so we should drop the connection.
+                self.close0(error: err, mode: .all, promise: nil)
+            }
+        default:
+            try self.setChannelSpecificOption0(option: option, value: value)
+        }
+    }
+
+    public func getOption<Option: ChannelOption>(_ option: Option) -> EventLoopFuture<Option.Value> {
+        if self.eventLoop.inEventLoop {
+            return self.eventLoop.makeCompletedFuture(Result { try getOption0(option: option) })
+        } else {
+            return eventLoop.submit { try self.getOption0(option: option) }
+        }
+    }
+
+    func getOption0<Option: ChannelOption>(option: Option) throws -> Option.Value {
+        self.eventLoop.assertInEventLoop()
+
+        guard !self.closed else {
+            throw ChannelError.ioOnClosedChannel
+        }
+
+        switch option {
+        case _ as ChannelOptions.Types.AutoReadOption:
+            return self.options.autoRead as! Option.Value
+        case _ as ChannelOptions.Types.SocketOption:
+            let optionValue = option as! ChannelOptions.Types.SocketOption
+
+            // SO_REUSEADDR and SO_REUSEPORT are handled here.
+            switch (optionValue.level, optionValue.name) {
+            case (SOL_SOCKET, SO_REUSEADDR):
+                return Int32(self.reuseAddress ? 1 : 0) as! Option.Value
+            case (SOL_SOCKET, SO_REUSEPORT):
+                return Int32(self.reusePort ? 1 : 0) as! Option.Value
+            default:
+                return try self.nwOptions.valueFor(socketOption: optionValue) as! Option.Value
+            }
+        case _ as ChannelOptions.Types.WriteBufferWaterMarkOption:
+            return self._backpressureManager.waterMarks as! Option.Value
+        case is NIOTSChannelOptions.Types.NIOTSEnablePeerToPeerOption:
+            return self.enablePeerToPeer as! Option.Value
+        case is NIOTSChannelOptions.Types.NIOTSAllowLocalEndpointReuse:
+            return self.allowLocalEndpointReuse as! Option.Value
```

Comment:
Why is this split, with `StateManagedNWConnectionChannel` handling the `get` but not the `set`?